src/TorchSharp/NN/Optimizer.cs

```diff
@@ -293,563 +452,1901 @@ public static AdadeltaOptimizer Adadelta(IEnumerable<(string name, Modules.Param
+                if (weight_decay < 0.0) throw new ArgumentException($"Invalid weight_decay value: {weight_decay}");
+                if (nesterov && (momentum <= 0 || dampening != 0)) throw new ArgumentException("Nesterov momentum requires a momentum and zero dampening");
+
+                var options = new Options {
+                    LearningRate = lr,
+                    InitialLearningRate = lr,
+                    dampening = dampening,
+                    maximize = maximize,
+                    momentum = momentum,
+                    nesterov = nesterov,
+                    weight_decay = weight_decay
+                };
+
+                _defaults = options;
+                _parameter_groups = new List<Modules.ParamGroup>();
+
+                foreach (var g in parameters) {
+                    add_param_group(g);
+                }
+            }
+
+            public override Tensor step(Func<Tensor> closure = null)
+            {
+                return _step<ParamGroup>(group => {
+
+                    var options = group.Options;
+                    var momentum = options.momentum.Value;
+                    var dampening = options.dampening.Value;
+                    var weight_decay = options.weight_decay.Value;
+                    var nesterov = options.nesterov.Value;
+                    var maximize = options.maximize.Value;
+                    var lr = options.LearningRate.Value;
+
+                    foreach (var param in group.Parameters) {
+
+                        var state = _state[param.handle];
+
+                        var grad = param.grad();
+
+                        if (grad is null) continue;
+
+                        if (weight_decay != 0) {
+                            grad = grad.add(param, alpha: weight_decay);
+                        }
+
+                        if (momentum != 0) {
+                            var buf = state.momentum_buffer;
+
+                            if (buf is null) {
+                                buf = grad.clone().detach().DetatchFromDisposeScope();
+                                state.momentum_buffer = buf;
+                            } else {
+                                buf.mul_(momentum).add_(grad, alpha: (1 - dampening));
+                            }
+
+                            if (nesterov) {
+                                grad = grad.add(buf, alpha: momentum);
+                            } else {
+                                grad = buf;
+                            }
+
+                            state.momentum_buffer = buf;
+                        }
+
+                        var alpha = maximize ? lr : -lr;
+                        param.add_(grad, alpha: alpha);
+
+                    }
+                }, closure);
+            }
+
+            protected override void Dispose(bool disposing)
+            {
+                base.Dispose(disposing);
+                foreach (var (name, state) in _state) {
```

Comment:
NIT: you can do `_state.Values` instead of unpacking `KeyValuePair` here.

Also, where is `_state` defined? When reading this change sequentially, I encountered it a few times by now, but without seeing its definition before the first use I am not certain about its contents.